access/email/app.go

```diff
@@ -0,0 +1,483 @@
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// checkTeleportVersion checks that Teleport version is not lower than required
+func (a *App) checkTeleportVersion(ctx context.Context) (proto.PingResponse, error) {
+	log := logger.Get(ctx)
+	log.Debug("Checking Teleport server version")
+	pong, err := a.apiClient.WithCallOptions(grpc.WaitForReady(true)).Ping(ctx)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			return pong, trace.Wrap(err, "server version must be at least %s", minServerVersion)
+		}
+		log.Error("Unable to get Teleport server version")
+		return pong, trace.Wrap(err)
+	}
+	err = lib.AssertServerVersion(pong, minServerVersion)
+	return pong, trace.Wrap(err)
+}
+
+// onWatcherEvent processes new incoming access request
+func (a *App) onWatcherEvent(ctx context.Context, event types.Event) error {
+	if kind := event.Resource.GetKind(); kind != types.KindAccessRequest {
+		return trace.Errorf("unexpected kind %q", kind)
+	}
+	op := event.Type
+	reqID := event.Resource.GetName()
+	ctx, _ = logger.WithField(ctx, "request_id", reqID)
+
+	switch op {
+	case types.OpPut:
+		ctx, _ = logger.WithField(ctx, "request_op", "put")
+		req, ok := event.Resource.(types.AccessRequest)
+		if !ok {
+			return trace.Errorf("unexpected resource type %T", event.Resource)
+		}
+		ctx, log := logger.WithField(ctx, "request_state", req.GetState().String())
+
+		var err error
+		switch {
+		case req.GetState().IsPending():
+			err = a.onPendingRequest(ctx, req)
+		case req.GetState().IsApproved():
+			err = a.onResolvedRequest(ctx, req)
+		case req.GetState().IsDenied():
+			err = a.onResolvedRequest(ctx, req)
+		default:
+			log.WithField("event", event).Warn("Unknown request state")
+			return nil
+		}
+
+		if err != nil {
+			log.WithError(err).Errorf("Failed to process request")
+			return trace.Wrap(err)
+		}
+
+		return nil
+	case types.OpDelete:
+		ctx, log := logger.WithField(ctx, "request_op", "delete")
+
+		if err := a.onDeletedRequest(ctx, reqID); err != nil {
+			log.WithError(err).Errorf("Failed to process deleted request")
+			return trace.Wrap(err)
+		}
+		return nil
+	default:
+		return trace.BadParameter("unexpected event operation %s", op)
+	}
+}
+
+// onPendingRequest is called when an access request is created or reviewed (with thresholds > 1)
+func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
+	log := logger.Get(ctx)
+
+	reqID, reqData := NewRequestData(req)
+
+	isNew, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{RequestData: reqData}, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if isNew {
+		if recipients := a.getEmailRecipients(ctx, req.GetSuggestedReviewers()); len(recipients) > 0 {
+			if err := a.sendNewThreads(ctx, recipients, reqID, reqData); err != nil {
+				return trace.Wrap(err)
+			}
+		} else {
+			log.Warning("No recipients to send")
```

Comment:
```suggestion
			log.Warning("No recipients to send")
			return nil
```

WDYT? In a similar place of slack plugin I don't do it but now i think we can just exit here.

Reply:
Agreed.